File: demo_spark/src/main/java/wordCount.scala
/**
  * Created by m2lgsb on 21/04/2018 09:44
  *
  * Scala 版本的单词数统计应用（暂时不需要深究细节）
  */

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._

// 创建一个Scala版本的Spark Context
val conf = new SparkConf().setMaster("local").setAppName("WordCount")
val sc = new SparkContext(conf)

// 读取我们的输入数据
val input = sc.textFile("./README.txt")

// 把他们切分成一个个单词
val words = input.flatMap(line => line.split(" "))

// 转换为键值对并计数
val counts = words.map(word => (word, 1)).reduceByKey{case (x, y) => x + y}
counts.saveAsTextFile("./output.txt")

File: demo_spark/src/main/java/wordCount.java
/**
 * Created by m2lgsb on 21/04/2018 12:34
 *
 * Java 版本的单词数统计应用（暂时不需要深究细节）
 */

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

public class wordCount {
    // 创建一个Java版本的Spark Context
    SparkConf conf = new SparkConf().setAppName("wordCount");
    JavaSparkContext sc = new JavaSparkContext(conf);

    // 读取我们的输入数据
    JavaRDD<String> input = sc.textFile("./README.txt");

    // 切分为单词
    JavaRDD<String> words = input.flatMap(
            new FlatMapFunction<String, String>() {
                public Iterator<String> call(String s) throws Exception {
                    return Arrays.asList(s.split(" "));
                }
            }
    );

    // 转换为键值对并计数
    JavaPairRDD<String, Integer> counts = words.mapToPair(
            new PairFunction<String, String, Integer>() {
                public Tuple2<String, Integer> call(String s) throws Exception {
                    return new Tuple2<String, Integer>(s, 1);
                }
            }
    ).reduceByKey(new Function2<Integer, Integer, Integer>() {
        public Integer call(Integer integer, Integer integer2) throws Exception {
            return integer + integer2;
        }
    });

    // 将统计出来的单词总数存入一个文本文件，引发求值
    counts.saveAsTextFile("./output/txt");
}
